src/corefx/System.Globalization.Native/collation.cpp

```diff
@@ -19,31 +19,75 @@ const int32_t CompareOptionsIgnoreCase = 1;
 // const int32_t CompareOptionsStringSort = 0x20000000;
 
 /*
+ * For increased performance, we cache the UCollator objects for a locale and
+ * share them across threads. This is safe (and supported in ICU) if we ensure
+ * multiple threads are only ever dealing with const UCollators.
+ */
+typedef struct _sort_handle
+{
+    UCollator* regular;
+    UCollator* ignoreCase;
+} SortHandle;
+
+/*
  * To collator returned by this function is owned by the callee and must be
- *closed when this method returns
- * with a U_SUCCESS UErrorCode.
+ * closed when this method returns with a U_SUCCESS UErrorCode.
  *
  * On error, the return value is undefined.
  */
-UCollator* GetCollatorForLocaleAndOptions(const char* lpLocaleName, int32_t options, UErrorCode* pErr)
+UCollator* CloneCollatorWithOptions(const UCollator* pCollator, int32_t options, UErrorCode* pErr)
 {
-    UCollator* pColl = nullptr;
+    UCollator* pClonedCollator = ucol_safeClone(pCollator, nullptr, nullptr, pErr);
 
-    pColl = ucol_open(lpLocaleName, pErr);
+    if ((options & CompareOptionsIgnoreCase) == CompareOptionsIgnoreCase)
+    {
+        ucol_setAttribute(pClonedCollator, UCOL_STRENGTH, UCOL_SECONDARY, pErr);
+    }
 
+    return pClonedCollator;
+}
+
+extern "C" SortHandle* GetSortHandle(const char* lpLocaleName)
+{
+    SortHandle* pSortHandle = (SortHandle*) malloc(sizeof(SortHandle));
+
+    UErrorCode err = U_ZERO_ERROR;
+
+    if (pSortHandle)
+    {
+        pSortHandle->regular = ucol_open(lpLocaleName, &err);
+        pSortHandle->ignoreCase = CloneCollatorWithOptions(pSortHandle->regular, CompareOptionsIgnoreCase, &err);
+    }
+
+    return U_SUCCESS(err) && (pSortHandle != nullptr) ? pSortHandle : nullptr;
+}
+
+extern "C" void CloseSortHandle(SortHandle* pSortHandle)
+{
+    ucol_close(pSortHandle->regular);
+    ucol_close(pSortHandle->ignoreCase);
+
+    pSortHandle->regular = nullptr;
+    pSortHandle->ignoreCase = nullptr;
+
+    free(pSortHandle);
+}
+
+const UCollator* GetCollatorFromSortHandle(const SortHandle* pSortHandle, int32_t options, UErrorCode* pErr)
+{
     if ((options & CompareOptionsIgnoreCase) == CompareOptionsIgnoreCase)
     {
-        ucol_setAttribute(pColl, UCOL_STRENGTH, UCOL_SECONDARY, pErr);
+        return pSortHandle->ignoreCase;
     }
 
-    return pColl;
+    return pSortHandle->regular;
 }
 
 /*
 Function:
 CompareString
 */
-extern "C" int32_t CompareString(const char* lpLocaleName,
+extern "C" int32_t CompareString(const SortHandle* pCollator,
```

Comment:
I think 'pCollator' would be better named as 'pSortHandle' as to not confuse pCollator and pColl inside the method.